prow/plugins/lint/lint.go

```diff
@@ -0,0 +1,226 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lint
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/golang/lint"
+
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const pluginName = "lint"
+
+var lintRe = regexp.MustCompile(`(?mi)^/lint\s*$`)
+
+func init() {
+	plugins.RegisterIssueCommentHandler(pluginName, handleIC)
+}
+
+type githubClient interface {
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+	CreateComment(org, repo string, number int, body string) error
+}
+
+func handleIC(pc plugins.PluginClient, ic github.IssueCommentEvent) error {
+	return handle(pc.GitHubClient, pc.GitClient, pc.Logger, ic)
+}
+
+// modifiedGoFiles returns a map from filename to patch string for all go files
+// that are modified in the PR.
+func modifiedGoFiles(ghc githubClient, org, repo string, number int) (map[string]string, error) {
+	changes, err := ghc.GetPullRequestChanges(org, repo, number)
+	if err != nil {
+		return nil, err
+	}
+
+	modifiedFiles := make(map[string]string)
+	for _, change := range changes {
+		if filepath.Ext(change.Filename) == ".go" {
+			modifiedFiles[change.Filename] = change.Patch
+		}
+	}
+	return modifiedFiles, nil
+}
+
+// problemsInPatches runs golint on the files. It returns all lint problems
+// that are in added lines in the files' patches.
+func problemsInFiles(r *git.Repo, files map[string]string) ([]lint.Problem, error) {
+	var problems []lint.Problem
+	l := new(lint.Linter)
+	for f, patch := range files {
+		src, err := ioutil.ReadFile(filepath.Join(r.Dir, f))
+		if err != nil {
+			return nil, err
+		}
+		ps, err := l.Lint(f, src)
+		if err != nil {
+			return nil, err
+		}
+		al, err := addedLines(patch)
+		if err != nil {
+			return nil, err
+		}
+		for _, p := range ps {
+			if al[p.Position.Line] {
+				problems = append(problems, p)
+			}
+		}
+	}
+	return problems, nil
+}
+
+func handle(ghc githubClient, gc *git.Client, log *logrus.Entry, ic github.IssueCommentEvent) error {
+	// Only handle open PRs and new requests.
+	if ic.Issue.State != "open" || !ic.Issue.IsPullRequest() || ic.Action != "created" {
+		return nil
+	}
+	if !lintRe.MatchString(ic.Comment.Body) {
+		return nil
+	}
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+
+	// List modified files.
+	modifiedFiles, err := modifiedGoFiles(ghc, org, repo, ic.Issue.Number)
```

Comment:
I don't mind spending one API token for a neat description of modified files and their patches. I could use the log but it's going to be more difficult.